src/twisted/internet/defer.py

```diff
@@ -160,6 +166,65 @@ def maybeDeferred(f, *args, **kw):
 
 
 
+def unbox(deferred, raises=True, passthrough=True):
+    """
+    Unbox a L{Deferred} that has already fired its result.
+
+    Take a deferred that has already fired its result and unbox it, returning
+    the result of the L{Deferred}. If the L{Deferred} is in an error state then
+    it will re-raise the original exception.
+
+    B{This function CANNOT take a Deferred with pending asynchronous work to be
+    done and make it synchronous. It can ONLY unbox a Deferred which already
+    has fired.}
+
+    @type deferred: Deferred
+    @param deferred: The Deferred to unbox.
+
+    @type raises: bool
+    @param raises: Whether or not to raise the original exception (L{True}) or
+        to return the failure object (L{False}).
+
+    @type passthrough: bool
+    @param passthrough: Whether or not to pass through the original value
+        (L{True}) or to consume any results, error or otherwise (L{False}).
+
+    @raise UnboxingWithoutResult: If C{deferred} has pending work to be done
+        and has no result to unbox.
+
+    @return: The result of the L{Deferred}.
+    """
+    results = []
+
+    @deferred.addBoth
+    def done(result):
+        results[:] = [result]
```

Comment:
This is _purely_ subjective, so please do with this feedback as you please, but I find the slicing/assigning idiom for extracting nonlocal variables to be somewhat hard to read, because you have to realize that the list isn't really a list, it's a holder.

Instead, I've started doing this:

```python
def done(result):
    done.result = result
nothing = object()
done.result = nothing
deferred.addBoth(done)
if done.result is not nothing:
    doSomethingWith(done.result)
else:
    "..."
```